numpyro/contrib/module.py

```diff
@@ -566,3 +566,169 @@ def __call__(self, x):
+    """
+    A primitive to create a random :mod:`equinox` style neural network
+    which can be used in MCMC samplers. The parameters of the neural network
+    will be sampled from ``prior``.
+
+    For supplying a prior dictionary, the dictionary keys are based on their jax key path. In equinox,
+    the key path may often start with a '.' as the first character - this is ignored in the keypath
+    here. To see the jax key paths for all of the leaves in your pytree model, you can run:
+
+        key_paths = [jtu.keystr(path) for path, _ in jtu.tree_leaves_with_path(model_instance)]
+
+    :param str name: name of the module to be registered.
+    :param eqx.Module nn_module: a pre-initialized `equinox` Module instance.
+    :param prior: a distribution or a dict of distributions or a callable.
+        If it is a distribution, all parameters will be sampled from the same
+        distribution. If it is a dict, it maps parameter names to distributions.
+        If it is a callable, it takes parameter name and parameter shape as
+        inputs and returns a distribution. For example::
+
+            class Linear(eqx.Module):
+                weight: jax.Array
+                bias: jax.Array
+
+                def __init__(self, in_size, out_size, key):
+                    wkey, bkey = jax.random.split(key)
+                    self.weight = jax.random.normal(wkey, (out_size, in_size))
+                    self.bias = jax.random.normal(bkey, (out_size,))
+
+                def __call__(self, x):
+                    return self.weight @ x + self.bias
+
+            # Eager initialization
+            linear = Linear(in_features=3, out_features=1, key=random.PRNGKey(0))
+            nn_priors = {"weight": dist.Normal(), "bias": dist.Cauchy()}
+            net = random_eqx_module("net", linear, prior=nn_priors)
+
+        Alternatively, we can use a callable. For example the following are equivalent::
+
+            prior=(lambda name, shape: dist.Cauchy() if name == 'bias' else dist.Normal())
+            prior={"weight": dist.Normal(), "bias": dist.Cauchy()}
+
+    :return: a callable that takes an array as an input and returns
+        the neural network transformed output array.
+    """
+    try:
+        import equinox as eqx
+    except ImportError as e:
+        raise ImportError(
+            "Looking like you want to use equinox to declare "
+            "nn modules. This is an experimental feature. "
+            "You need to install the latest version of `equinox` to use this feature. "
+            "It can be installed with `pip install git+https://github.com/patrick-kidger/equinox.git`."
+        ) from e
+
+    nn = eqx_module(name, nn_module)
+    params, static = eqx.partition(nn, filter_spec=eqx.is_inexact_array)
+    with numpyro.handlers.scope(prefix=name):
+        new_params = update_tree_params(params, prior=prior)
+
+    return eqx.combine(new_params, static)
+
+
+def update_tree_params(params, prior):
```

Comment:
Can we use update_params like in flax/nnx modules to make it easier to maintain the logic and to make things consistent?

Reply:
Thats a good question, but I'm not sure how possible it is. I originally tried to do this, but unlike `nnx` there arent any methods for going back and forth between equinox pytrees and dictionaries (as far as I know)

I'm not a pytree expert so if you happen to know if this capability exists in jax  or equinox then I could probably use `_update_params` like the other modules do

I used this equinox issue as a reference https://redirect.github.com/patrick-kidger/equinox/issues/657
.

Reply:
@fehiepsi do you have any more thoughts here? equinox doesnt have tools to go back and forth between dictionaries and pytrees like nnx does

Reply:
alright just updated this to work with dictionaries and utilize the original _update_params() function thanks to Patricks help!